sage.py

```diff
@@ -481,483 +449,311 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-                    timestamps = []
-                    
-                    for e in mindata:
-                        if len(e) > 0:
-                            raw_ports.append([(x[3]) for x in e])
-                            raw_sign.append([(x[4]) for x in e])
-                            timestamps.append([(x[2]) for x in e])
-                        else:
-                            raw_ports.append([])
-                            raw_sign.append([])
-                            timestamps.append([])
+            _team_times['->'.join(attacker_victim)] = first_elapsed_time
 
-                    _flat_ports = [item for sublist in raw_ports for item in sublist]
-                    
-                    # The following makes exact start/end times based on alert timestamps
-                    filtered_timestamps = [timestamps[x[0]:x[1]+1] for x in episodes]
-                    start_end_timestamps = [(sorted([item for sublist in x for item in sublist])[0], sorted([item for sublist in x for item in sublist])[-1]) for x in filtered_timestamps]
-                    minute_info = [((x[0]-startTimes[tid]).total_seconds(), (x[1]-startTimes[tid]).total_seconds()) for x in start_end_timestamps]
-
-                    episode = [(mi[0], mi[1], mcat, events[x[0]:x[1]+1], 
-                                   raw_ports[x[0]:x[1]+1], raw_sign[x[0]:x[1]+1], filtered_timestamps)
-                                 for x,mi in zip(episodes, minute_info)] # now the start and end are actual elapsed times
-                    
-                    #EPISODE DEF: (startTime, endTime, mcat, len(rawevents), volume(alerts), epiPeriod, epiServices, list of unique signatures, (1st timestamp, last timestamp)
-                    episode = [(x[0], x[1], x[2], x[3], round(sum(x[3])/float(len(x[3])),1), (x[1]-x[0]),
-                                [item for sublist in x[4] for item in sublist], list(set([item for sublist in x[5] for item in sublist])),
-                                se_ts) for x,se_ts in zip(episode,start_end_timestamps)]
-                    h_ep.extend(episode) 
+            ts = [x[2] for x in alerts]
+            diff_ts = [0.0]
+            for i in range(1, len(ts)):
+                diff_ts.append(round((ts[i] - ts[i - 1]).total_seconds(), 2))
+            elapsed_time = list(accumulate(diff_ts))
+            relative_elapsed_time = [round(x + first_elapsed_time, 2) for x in elapsed_time]
 
-            if len(h_ep) == 0:
+            host_episodes = []
+            for mcat in mcats:
+                # 2.5-minute (150s) fixed step (window). Can be reduced or increased depending on required granularity
+                alert_seq = []
+                for i in range(int(first_elapsed_time), int(relative_elapsed_time[-1]), step):
+                    window = [a for dt, a in zip(relative_elapsed_time, alerts) if (i <= dt < (i + step)) and a[1] == mcat]
+                    alert_seq.append(window)  # Alerts per 'step' seconds (window)
```

Comment:
From what I have understood, this is essentially the alert sequence that was mentioned in the paper, so I have renamed it accordingly (I don't know what the `mindata` referred to)

Reply:
It was short for minute_data, i.e., how many alerts within a single 2.5-minute window were triggered. Perhaps a better name for this would be hyperalert_seq (since each element captures multiple alerts)

Reply:
Hmm, that's if fine, but in the paper, `alert sequence` was explicitly defined as a windowed sequence, which is why I used this name (for consistency). Do you want me to rename it to `hyperalert_seq`?

![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/d27c8a1d-92b9-4132-8656-5d628904fa27)

Reply:
Updated, as discussed in the meeting